craft_providers/lxd/lxd_instance.py

```diff
@@ -508,33 +508,79 @@ def push_file(self, *, source: pathlib.Path, destination: pathlib.PurePath) -> N
         )
 
     def start(self) -> None:
-        """Start instance.
+        """Start the instance.
 
-        :raises LXDError: on unexpected error.
+        :raises LXDError: If the instance fails to start.
         """
         logger.info("Starting instance")
         self.lxc.start(
             instance_name=self.instance_name, project=self.project, remote=self.remote
         )
 
+        def _is_running(timeout: float) -> None:
+            """Raise an error if the instance isn't running."""
+            if self.is_running():
+                return
+            raise LXDError("Instance isn't running")
+
+        # `lxc start` is an asynchronous operation, so wait until the instance
+        # starts before returning.
+        retry.retry_until_timeout(
+            timeout=TIMEOUT_SIMPLE,
+            retry_wait=RETRY_WAIT,
+            func=_is_running,
+            error=LXDError(brief="Instance failed to start."),
+        )
+
     def restart(self) -> None:
-        """Restart instance.
+        """Restart the instance.
 
-        :raises LXDError: on unexpected error.
+        :raises LXDError: If the instance fails to restart.
         """
         self.lxc.restart(
             instance_name=self.instance_name, project=self.project, remote=self.remote
         )
 
+        def _is_running(timeout: float) -> None:
+            """Raise an error if the instance isn't running."""
+            if self.is_running():
+                return
+            raise LXDError(brief="Instance isn't running.")
+
+        # `lxc restart` is an asynchronous operation, so wait until the instance
+        # restarts before returning.
+        retry.retry_until_timeout(
+            timeout=TIMEOUT_SIMPLE,
+            retry_wait=RETRY_WAIT,
+            func=_is_running,
+            error=LXDError(brief="Instance failed to restart."),
+        )
+
     def stop(self) -> None:
-        """Stop instance.
+        """Stop the instance.
 
-        :raises LXDError: on unexpected error.
+        :raises LXDError: If the instance fails to stop.
         """
         self.lxc.stop(
             instance_name=self.instance_name, project=self.project, remote=self.remote
         )
 
+        def _is_stopped(timeout: float) -> None:
```

Comment:
Every time we do more work here it looks more and more like something that would fit so naturally with `asyncio`...